File: ABC-226/C.cpp
#include <iostream>
#include <vector>

using namespace std;
using Graph = vector<vector<int>>;

vector<bool> seen;

void dfs(const Graph &G, int v) {
  seen[v] = true;
  for (auto next_v : G[v]) {
    if (seen[next_v])
      continue;
    dfs(G, next_v);
  }
}

int main() {
  int n;
  cin >> n;
  Graph G(n);
  vector<int> T;

  for (int i = 0; i < n; i++) {
    int t, k;
    cin >> t >> k;
    T.push_back(t);
    for (int j = 0; j < k; j++) {
      int a;
      cin >> a;
      a--;
      G[i].push_back(a);
    }
  }

  seen.assign(n, false);
  dfs(G, n - 1);

  long long ans = 0;
  for (int i = 0; i < n; i++) {
    if (seen[i]) {
      ans += T[i];
    }
  }

  cout<<ans<<endl;

  return 0;
}

File: ABC-226/C.py
import sys

sys.setrecursionlimit(5000)
n = int(input())

G = [[] for _ in range(n)]

T = []
for i in range(n):
    line = list(map(int, input().split()))
    t = line[0]
    T.append(t)
    k = line[1]
    l = [a - 1 for a in line[2 : 2 + k]]
    G[i] += l


seen = [False] * n
waza = set()


def dfs(G, v):
    seen[v] = True
    waza.add(v)

    for i in G[v]:
        if seen[i]:
            continue
        dfs(G, i)


dfs(G, n - 1)
waza = list(waza)

ans = 0
for i in range(n):
    if i in waza:
        ans += T[i]

print(ans)
